<!-- 
RSS generated by JIRA (6.4.11#64026-sha1:78f6ec473a3f058bd5d6c30e9319c7ab376bdb9c) at Thu Dec 22 11:53:47 UTC 2016

It is possible to restrict the fields that are returned in this document by specifying the 'field' parameter in your request.
For example, to request only the issue key and summary add field=key&field=summary to the URL of your request.
For example:
https://jira.spring.io/si/jira.issueviews:issue-xml/BATCH-1572/BATCH-1572.xml?field=key&amp;field=summary
-->
<rss version="0.92" >
<channel>
    <title>Spring JIRA</title>
    <link>https://jira.spring.io</link>
    <description>This file is an XML representation of an issue</description>
    <language>en-uk</language>    <build-info>
        <version>6.4.11</version>
        <build-number>64026</build-number>
        <build-date>25-08-2015</build-date>
    </build-info>

<item>
            <title>[BATCH-1572] Step not failing on org.springframework.transaction.UnexpectedRollbackException</title>
                <link>https://jira.spring.io/browse/BATCH-1572</link>
                <project id="10090" key="BATCH">Spring Batch</project>
                    <description>&lt;p&gt;The transaction is timing out for a read by the reader in a step due to a lock. The transaction manager marks the out come of the transaction to be rollback only. Once the lock is released and the reader returns  a org.springframework.transaction.UnexpectedRollbackException is thrown while updating the chunk to database. Batch is considering this exception to be a non fatal exception and is continuing with the next chunk just by logging the exception at a debug level. Instead ity should fail the step I guess.&lt;/p&gt;

&lt;p&gt;Problem : I am loosing all the records in that chunk. &lt;/p&gt;

&lt;p&gt;I changed the org.springframework.batch.core.step.item.SimpleRetryExceptionHandler.java&apos;s constructor like &lt;/p&gt;

&lt;p&gt;public SimpleRetryExceptionHandler(RetryPolicy retryPolicy, ExceptionHandler exceptionHandler, Collection&amp;lt;Class&amp;lt;? extends Throwable&amp;gt;&amp;gt; fatalExceptionClasses) &lt;/p&gt;
{
		this.retryPolicy = retryPolicy;
		this.exceptionHandler = exceptionHandler;
		fatalExceptionClasses.add(org.springframework.transaction.UnexpectedRollbackException.class);
		this.fatalExceptionClassifier = new BinaryExceptionClassifier(fatalExceptionClasses);
	}

&lt;p&gt;and tested. Now it is failing the step.&lt;/p&gt;


</description>
                <environment>windows xp, windows 2003 sp2, jotm jta transaction manager, sql server 2005</environment>
        <key id="31440">BATCH-1572</key>
            <summary>Step not failing on org.springframework.transaction.UnexpectedRollbackException</summary>
                <type id="1" iconUrl="https://jira.spring.io/images/icons/issuetypes/bug.png">Bug</type>
                                            <priority id="3" iconUrl="https://jira.spring.io/images/icons/priorities/major.png">Major</priority>
                        <status id="5" iconUrl="https://jira.spring.io/images/icons/statuses/resolved.png" description="A resolution has been taken, and it is awaiting verification by reporter. From here issues are either reopened, or are closed.">Resolved</status>
                    <statusCategory id="3" key="done" colorName="green"/>
                                    <resolution id="8">Complete</resolution>
                                        <assignee username="david_syer">Dave Syer</assignee>
                                    <reporter username="springilu">Srinivas Arava</reporter>
                        <labels>
                    </labels>
                <created>Thu, 20 May 2010 02:20:50 +0000</created>
                <updated>Mon, 27 Sep 2010 01:17:31 +0000</updated>
                            <resolved>Mon, 27 Sep 2010 01:17:31 +0000</resolved>
                                    <version>2.1.1</version>
                                    <fixVersion>2.1.4</fixVersion>
                                    <component>Core</component>
                        <due></due>
                            <votes>0</votes>
                                    <watches>1</watches>
                                    <timeoriginalestimate seconds="0">0d</timeoriginalestimate>
                            <timeestimate seconds="0">0d</timeestimate>
                                        <comments>
                            <comment id="53985" author="springilu" created="Fri, 21 May 2010 08:46:35 +0000"  >&lt;p&gt;I attached a eclipse project interchange file here. It is a small example program to recreate the problem. It requires simple database setup. I used derby network server.&lt;/p&gt;</comment>
                            <comment id="57364" author="david_syer" created="Tue, 31 Aug 2010 05:22:24 +0000"  >&lt;p&gt;This turns out to be considerably more difficult than it may at first seem.  The problem is that when a commit fails (after all business processing has completed) we should not only fail the step, but also roll back the changes to the in-memory StepExecution that we are updating.  There is no way with existing implementations of step to recover the step execution data from before the transaction.  I will keep digging to see if there is anything that can be done in 2.1.x.&lt;/p&gt;

&lt;p&gt;Even if there was, a failed commit should still be a retryable exception, but there is no way to signal back to the chunk (which completed successfully) that it can retry the last transaction.  It would be nice to be able  to do that, to preserve retryability, but I guess that&apos;s a nice to have and could be relegated to a separate issue for 3.0.&lt;/p&gt;</comment>
                            <comment id="57674" author="david_syer" created="Sat, 4 Sep 2010 04:29:48 +0000"  >&lt;p&gt;Resolved the step execution rollback issue with an explicit copy of the relevant fields.  This is quite a big change, so here&apos;s hoping the unit tests are complete enough to detect errors.&lt;/p&gt;</comment>
                            <comment id="58481" author="david_syer" created="Mon, 27 Sep 2010 01:00:33 +0000"  >&lt;p&gt;Re-opening because restart is broken: we need to copy the execution context as well.&lt;/p&gt;</comment>
                            <comment id="58483" author="david_syer" created="Mon, 27 Sep 2010 01:17:30 +0000"  >&lt;p&gt;Fixed (again) by copying execution context as well.&lt;/p&gt;</comment>
                    </comments>
                    <attachments>
                            <attachment id="16581" name="test.zip" size="1752527" author="springilu" created="Fri, 21 May 2010 08:46:35 +0000"/>
                    </attachments>
                <subtasks>
                    </subtasks>
                <customfields>
                                                                            <customfield id="customfield_10170" key="com.atlassian.jira.ext.charting:firstresponsedate">
                        <customfieldname>First Response Date</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue>Tue, 31 Aug 2010 05:22:24 +0000</customfieldvalue>

                        </customfieldvalues>
                    </customfield>
                                                                                            <customfield id="customfield_10280" key="com.pyxis.greenhopper.jira:gh-global-rank">
                        <customfieldname>Global Rank</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue>11152</customfieldvalue>
                        </customfieldvalues>
                    </customfield>
                                                                                            <customfield id="customfield_10880" key="com.pyxis.greenhopper.jira:gh-lexo-rank">
                        <customfieldname>Rank</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue>0|i03v2v:</customfieldvalue>

                        </customfieldvalues>
                    </customfield>
                                                                <customfield id="customfield_10380" key="com.pyxis.greenhopper.jira:gh-global-rank">
                        <customfieldname>Rank (Obsolete)</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue>22577</customfieldvalue>
                        </customfieldvalues>
                    </customfield>
                                                                <customfield id="customfield_10381" key="com.pyxis.greenhopper.jira:gh-global-rank">
                        <customfieldname>Ranking</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue>19058</customfieldvalue>
                        </customfieldvalues>
                    </customfield>
                                                                <customfield id="customfield_10120" key="com.atlassian.jira.plugin.system.customfieldtypes:url">
                        <customfieldname>Reference URL</customfieldname>
                        <customfieldvalues>
                            <customfieldvalue><![CDATA[http://forum.springsource.org/showthread.php?t=89000]]></customfieldvalue>

                        </customfieldvalues>
                    </customfield>
                                                                                                                        <customfield id="customfield_10171" key="com.atlassian.jira.ext.charting:timeinstatus">
                        <customfieldname>Time in Status</customfieldname>
                        <customfieldvalues>
                            
                        </customfieldvalues>
                    </customfield>
                                                                </customfields>
    </item>
</channel>
</rss>